launchlib/launcher.go

```diff
@@ -33,38 +33,49 @@ const (
 	ExecPathBlackListRegex = `[^\w.\/_\-]`
 )
 
-type ServiceCommands struct {
-	Primary      *exec.Cmd
-	SubProcesses map[string]*exec.Cmd
+type ProcCmd struct {
+	Name   string
+	Cmd    *exec.Cmd
+	Stdout io.Writer
```

Comment:
This should probably not be here.  It is pretty bad practice to be returning things which need to be closed from anything but minimal helper functions as it immediately becomes very unclear who is responsible for closing them.  At the moment, it doesn't look like they get closed anywhere.

Reply:
Yikes, thanks for catching this.

Reply:
Ok, I'm changing everything to not return open files so it's clear who owns their closing, but in doing so, I'm seeing that it seems like we can't even close these files anyway. Each is opened right before we set each `Cmd`'s `Stdout` to it. Then our program starts the `Cmd` and then exits and the child process gets adopted by `init`. The process image is completely replaced by whatever command we started. It seems the only way to ensure these files' closure is to let the OS do it when the process terminates.

Reply:
You can close them after starting the command.  Under the hood, starting the command forks the current process which creates a copy of the parents file descriptors, we in the parent can then close the file, since the child already has it's own open copy.

Reply:
Ah, right, I forgot that (I was wondering how the parent could close the file and allow the child to still read it).

This would be a good `./godelw verify` check!